pyembed/src/pyalloc.rs

```diff
@@ -150,6 +210,141 @@ pub fn make_raw_rust_memory_allocator() -> RawAllocator {
     }
 }
 
+pub fn make_rust_pyobject_arena_allocator() -> PyAllocator {
+    // We need to allocate the HashMap on the heap so the pointer doesn't refer
+    // to the stack. We rebox and add the Box to our struct so lifetimes are
+    // managed.
+    let alloc = Box::new(HashMap::<*mut u8, alloc::Layout>::new());
+    let state = Box::into_raw(alloc);
+
+    let allocator = pyffi::PyObjectArenaAllocator {
+        ctx: state as *mut c_void,
+        alloc: Some(rust_pymalloc),
+        free: Some(rust_pymallocfree),
+    };
+
+    PyAllocator {
+        allocator,
+        _state: unsafe { Box::from_raw(state) },
+    }
+}
+// Now let's define a raw memory allocator that interfaces directly with mimalloc.
+// This avoids the overhead of going through Rust's allocation layer.
+#[cfg(feature = "mimalloc")]
+extern "C" fn raw_mimalloc_malloc(_ctx: *mut c_void, size: size_t) -> *mut c_void {
+    // PyMem_RawMalloc()'s docs say: Requesting zero bytes returns a distinct
+    // non-NULL pointer if possible, as if PyMem_RawMalloc(1) had been called
+    // instead.
+    let size = match size {
+        0 => 1,
+        val => val,
+    };
+    // Allocate `size` bytes.Returns pointer to the allocated memory or null if out of memory.
+    // Returns a unique pointer if called with `size` 0
+	let padding = _ctx as *const _ as usize;
+    unsafe { mimallocffi::mi_malloc(size + padding) }
```

Comment:
The padding was based on the author victor Steiner explanation of the custom memory allocators. Apparently ctx is used in Python for padding 10 mb or kB for pymemallocatorex and 2 mb or kB for pymalloc. However, my recent tests have shown this can max the memory allocation too much and I’ve decided to not use it on non-heap allocator version. The heap allocator I’m working on it works nicely with the padding.

so as mentioned earlier we can remove the padding here for non-heap mimalloc, which I’ve done locally.

Reply:
Stinner* here is his explanation: https://www.python.org/dev/peps/pep-0445/

again I don’t recommend until we implement the mimalloc heap version though

Reply:
OK. I was confused because we don't actually set `ctx` to non-NULL in this patch. Thanks for the explanation!